src/main/java/net/starlark/java/lib/toml/TomlParser.java

```diff
@@ -0,0 +1,567 @@
+import net.starlark.java.eval.Starlark;
+import net.starlark.java.eval.StarlarkCallable;
+import net.starlark.java.eval.StarlarkFloat;
+import net.starlark.java.eval.StarlarkInt;
+import net.starlark.java.eval.StarlarkIterable;
+import net.starlark.java.eval.StarlarkList;
+import net.starlark.java.eval.StarlarkSemantics;
+import net.starlark.java.eval.StarlarkSet;
+import net.starlark.java.eval.StarlarkThread;
+import net.starlark.java.eval.StarlarkValue;
+import net.starlark.java.eval.Structure;
+import net.starlark.java.lib.json.Json;
+import org.tomlj.Toml;
+import org.tomlj.TomlArray;
+import org.tomlj.TomlParseResult;
+import org.tomlj.TomlTable;
+
+// Tests at //src/test/java/net/starlark/java/eval:testdata/toml.star
+
+/**
+ * TomlParser defines the Starlark {@code toml} module, which provides functions for encoding/decoding
+ */
+@StarlarkBuiltin(
+    name = "toml",
+    category = "core.lib",
+    doc = "Module toml is a Starlark module of TOML-related functions.")
+public final class TomlParser implements StarlarkValue {
+
+  private TomlParser() {}
+
+  private static final Pattern BARE_KEY = Pattern.compile("[A-Za-z0-9_-]+");
+
+  /**
+   * The module instance. You may wish to add this to your predeclared environment under the name
+   * "toml".
+   */
+  public static final TomlParser INSTANCE = new TomlParser();
+
+  /**
+   * Encodes a Starlark value as TOML.
+   */
+  @StarlarkMethod(
+      name = "encode",
+      doc =
+          "<p>The encode function accepts one required positional argument, which it converts to"
+              + " TOML by cases:\n"
+              + "<ul>\n"
+              + "<li>True, and False are converted to 'true', and 'false',"
+              + " respectively.\n"
+              + "<li>An int, no matter how large, is encoded as a decimal integer. Some decoders"
+              + " may not be able to decode very large integers.\n"
+              + "<li>A float is encoded using a decimal point or an exponent or both, even if its"
+              + " numeric value is an integer. It is an error to encode a non-finite "
+              + " floating-point value.\n"
+              + "<li>A string value is encoded as a TOML string literal that denotes the value.\n"
+              + "<li>A dict is encoded as a TOML table. It is an error if any key is not a string.\n"
+              + "<li>A list or tuple is encoded as a TOML array.\n"
+              + "</ul>\n"
+              + "Encoding any other value yields an error.\n",
+      parameters = {@Param(name = "x")},
+      useStarlarkThread = true)
+  public String encode(Object x, StarlarkThread thread) throws EvalException, InterruptedException {
+    try {
+      Object javaValue = convertToJava(x, thread.getSemantics());
+
+      // TOML requires the root to be a map/table structure
+      if (javaValue instanceof Map) {
+        StringBuilder sb = new StringBuilder();
+        encodeTable(sb, (Map<String, Object>)javaValue, new ArrayList<>());
```

Comment:
Since you already support encoding structs as toml tables - why not allow `x` to be a struct or NativeInfo?